File: backend/app/Models/RefundException.ts
import { DateTime } from 'luxon'
import { BaseModel, column, belongsTo, BelongsTo } from '@ioc:Adonis/Lucid/Orm'
import Order from './Order'
import Subscription from './Subscription'
import User from './User'

export default class RefundException extends BaseModel {
  @column({ isPrimary: true })
  public id: number

  @column()
  public orderId: number

  @column()
  public subscriptionId: number | null

  @column()
  public exceptionType: string

  @column()
  public status: string

  @column()
  public refundAmount: number

  @column()
  public reason: string | null

  @column()
  public description: string | null

  @column()
  public reportedBy: number | null

  @column()
  public handlerId: number | null

  @column()
  public handlerConclusion: string | null

  @column.dateTime()
  public handledAt: DateTime | null

  @column.dateTime({ autoCreate: true })
  public createdAt: DateTime

  @column.dateTime({ autoCreate: true, autoUpdate: true })
  public updatedAt: DateTime

  @belongsTo(() => Order, { foreignKey: 'orderId' })
  public order: BelongsTo<typeof Order>

  @belongsTo(() => Subscription, { foreignKey: 'subscriptionId' })
  public subscription: BelongsTo<typeof Subscription>

  @belongsTo(() => User, { foreignKey: 'reportedBy' })
  public reporter: BelongsTo<typeof User>

  @belongsTo(() => User, { foreignKey: 'handlerId' })
  public handler: BelongsTo<typeof User>
}

File: backend/app/models/refund_exception.py
from datetime import datetime
from sqlalchemy import Column, Integer, String, DateTime, Enum, ForeignKey, Text, Boolean
from sqlalchemy.orm import relationship

from app.core.database import Base
import enum


class RefundExceptionStatus(str, enum.Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    RESOLVED = "resolved"
    ESCALATED = "escalated"


class RefundException(Base):
    __tablename__ = "refund_exceptions"

    id = Column(Integer, primary_key=True, index=True)
    registration_id = Column(Integer, ForeignKey("registrations.id"), nullable=False, index=True)
    exception_type = Column(String(100))
    description = Column(Text)
    status = Column(Enum(RefundExceptionStatus), default=RefundExceptionStatus.PENDING, index=True)
    
    refund_amount = Column(Integer, default=0)
    actual_refund_amount = Column(Integer, default=0)
    
    handled_by = Column(String(100))
    handle_result = Column(Text)
    
    auto_generated = Column(Boolean, default=False)
    
    created_at = Column(DateTime, default=datetime.utcnow, index=True)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    resolved_at = Column(DateTime)

    registration = relationship("Registration", back_populates="refund_exceptions")
    todo = relationship("Todo", back_populates="refund_exception", uselist=False)
